components/backup-stream/src/utils.rs

```diff
@@ -589,6 +594,42 @@ pub fn is_overlapping(range: (&[u8], &[u8]), range2: (&[u8], &[u8])) -> bool {
     }
 }
 
+pub struct FilesReader {
+    files: Vec<Pin<Box<File>>>,
+    index: usize,
+}
+
+impl FilesReader {
+    pub fn new(files: Vec<Pin<Box<File>>>) -> Self {
+        FilesReader { files, index: 0 }
+    }
+}
+
+impl AsyncRead for FilesReader {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> Poll<std::io::Result<()>> {
+        let me = self.get_mut();
+
+        while me.index < me.files.len() {
+            let rem = buf.remaining();
+            match me.files[me.index].as_mut().poll_read(cx, buf) {
```

Comment:
... If `pool_read` needs `Pin<&mut Self>` here, perhaps we can pin it at stack by the [`pin!`](https://docs.rs/tokio/latest/tokio/macro.pin.html) macro.

Reply:
... Or just `Pin::new(&mut me.files[me.index])` if `pin!` needs move out.